File: merge_words/java/MergeWords.java
package merge_words.java;

import java.util.*;

public class MergeWords {
    private final List<String> dictionary = Arrays.asList("the", "cat", "dog", "hat", "rat", "bat", "mat", "sat", "fat", "pat", "tat", "eat", "tea", "ate", "eta");
    private List<List<String>> tiles;
    private List<Integer> selectedTiles;
    private int score;
    private List<String> foundWords;

    public MergeWords() {
        initializeGame();
    }

    public void initializeGame() {
        tiles = generateTiles();
        selectedTiles = new ArrayList<>();
        score = 0;
        foundWords = new ArrayList<>();
    }

    private List<List<String>> generateTiles() {
        List<List<String>> tileOptions = Arrays.asList(
            Arrays.asList("t", "h", "e"), Arrays.asList("c", "a"), List.of("t"),
            Arrays.asList("d", "o"), List.of("g"), Arrays.asList("h", "a"),
            Arrays.asList("t", "e"), Arrays.asList("c", "a", "t"), Arrays.asList("d", "o", "g")
        );
        Collections.shuffle(tileOptions);
        return new ArrayList<>(tileOptions.subList(0, 9));
    }

    public boolean selectTile(int index) {
        if (index >= 0 && index < tiles.size() && !selectedTiles.contains(index)) {
            selectedTiles.add(index);
            return true;
        }
        return false;
    }

    public boolean backspace() {
        if (!selectedTiles.isEmpty()) {
            selectedTiles.removeLast();
            return true;
        }
        return false;
    }

    public String getCurrentWord() {
        StringBuilder word = new StringBuilder();
        for (int i : selectedTiles) {
            word.append(String.join("", tiles.get(i)));
        }
        return word.toString();
    }

    public String submitWord() {
        String word = getCurrentWord();
        if (word.isEmpty()) {
            return "No tiles selected";
        }
        if (foundWords.contains(word)) {
            selectedTiles.clear();
            return "Word already used";
        }
        if (dictionary.contains(word.toLowerCase())) {
            score += word.length();
            foundWords.add(word);
            selectedTiles.clear();
            return "Valid word! Score: " + score;
        } else {
            selectedTiles.clear();
            return "Invalid word";
        }
    }

    public void shuffleTiles() {
        tiles = generateTiles();
        selectedTiles.clear();
    }

    public boolean isGameOver() {
        return generatePossibleWords().isEmpty();
    }

    private List<String> generatePossibleWords() {
        List<String> words = new ArrayList<>();
        for (int i = 1; i <= tiles.size(); i++) {
            for (List<Integer> combo : getCombinations(new ArrayList<>(), 0, i)) {
                StringBuilder word = new StringBuilder();
                for (int j : combo) {
                    word.append(String.join("", tiles.get(j)));
                }
                String wordStr = word.toString();
                if (dictionary.contains(wordStr.toLowerCase()) && !foundWords.contains(wordStr)) {
                    words.add(wordStr);
                }
            }
        }
        return words;
    }

    private List<List<Integer>> getCombinations(List<Integer> curr, int start, int k) {
        List<List<Integer>> results = new ArrayList<>();
        if (curr.size() == k) {
            results.add(new ArrayList<>(curr));
            return results;
        }
        for (int i = start; i < tiles.size(); i++) {
            curr.add(i);
            results.addAll(getCombinations(curr, i + 1, k));
            curr.removeLast();
        }
        return results;
    }

    public List<List<String>> getTiles() {
        return tiles;
    }

    public List<Integer> getSelectedTiles() {
        return selectedTiles;
    }

    public int getScore() {
        return score;
    }

    public List<String> getFoundWords() {
        return foundWords;
    }
}

File: merge_words/python/merge_words.py
#!/usr/bin/env python3

import random

class MergeWords:
    def __init__(self):
        self.dictionary = ["the", "cat", "dog", "hat", "rat", "bat", "mat", "sat", "fat", "pat", "tat", "eat", "tea", "ate", "eta"]
        self.tiles = []
        self.selected_tiles = []
        self.score = 0
        self.found_words = []
        self.initialize_game()

    def initialize_game(self):
        self.tiles = self.generate_tiles()
        self.selected_tiles = []
        self.score = 0
        self.found_words = []

    def generate_tiles(self):
        tile_options = [
            ['t', 'h', 'e'], ['c', 'a'], ['t'], ['d', 'o'], ['g'],
            ['h', 'a'], ['t', 'e'], ['c', 'a', 't'], ['d', 'o', 'g']
        ]
        random.shuffle(tile_options)
        return tile_options[:9]

    def select_tile(self, index):
        if 0 <= index < len(self.tiles) and index not in self.selected_tiles:
            self.selected_tiles.append(index)
            return True
        return False

    def backspace(self):
        if self.selected_tiles:
            self.selected_tiles.pop()
            return True
        return False

    def get_current_word(self):
        return ''.join(''.join(self.tiles[i]) for i in self.selected_tiles)

    def submit_word(self):
        word = self.get_current_word()
        if not word:
            return False, "No tiles selected"
        if word in self.found_words:
            self.selected_tiles = []
            return False, "Word already used"
        if word.lower() in self.dictionary:
            self.score += len(word)
            self.found_words.append(word)
            self.selected_tiles = []
            return True, f"Valid word! Score: {self.score}"
        else:
            self.selected_tiles = []
            return False, "Invalid word"

    def shuffle_tiles(self):
        self.tiles = self.generate_tiles()
        self.selected_tiles = []

    def is_game_over(self):
        return len(self.generate_possible_words()) == 0

    def generate_possible_words(self):
        words = []
        for i in range(1, len(self.tiles) + 1):
            for combo in self.get_combinations(list(range(len(self.tiles))), i):
                word = ''.join(''.join(self.tiles[j]) for j in combo)
                if word.lower() in self.dictionary and word not in self.found_words:
                    words.append(word)
        return words

    def get_combinations(self, arr, k):
        results = []
        def combine(curr, start, k):
            if len(curr) == k:
                results.append(curr[:])
                return
            for i in range(start, len(arr)):
                curr.append(arr[i])
                combine(curr, i + 1, k)
                curr.pop()
        combine([], 0, k)
        return results